kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/CheckpointV2ReadSuite.scala

```diff
@@ -0,0 +1,291 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults
+
+import java.io.File
+
+import scala.collection.JavaConverters._
+
+import io.delta.kernel.defaults.utils.{TestRow, TestUtils}
+import io.delta.kernel.internal.{InternalScanFileUtils, SnapshotImpl}
+import io.delta.tables.DeltaTable
+import org.apache.hadoop.fs.Path
+import org.scalatest.funsuite.AnyFunSuite
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.delta.{DeltaLog, Snapshot}
+import org.apache.spark.sql.delta.actions.{AddFile, Metadata, Protocol}
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.delta.test.DeltaTestImplicits.OptimisticTxnTestHelper
+import org.apache.spark.sql.delta.util.FileNames
+import org.apache.spark.sql.types.{BooleanType, IntegerType, LongType, MapType, StringType, StructType}
+
+class CheckpointV2ReadSuite extends AnyFunSuite with TestUtils {
+  private final val supportedFileFormats = Seq("json", "parquet")
+
+  def createSourceTable(tbl: String, path: String): Unit = {
+    spark.sql(s"CREATE TABLE $tbl (a INT, b STRING) USING delta CLUSTER BY (a) " +
+      s"LOCATION '$path' " +
+      s"TBLPROPERTIES ('delta.checkpointInterval' = '2', 'delta.checkpointPolicy'='v2')")
+    spark.sql(s"INSERT INTO $tbl VALUES (1, 'a'), (2, 'b')")
+    spark.sql(s"INSERT INTO $tbl VALUES (3, 'c'), (4, 'd')")
+    spark.sql(s"INSERT INTO $tbl VALUES (5, 'e'), (6, 'f')")
+  }
+
+  def validateSnapshot(
+      path: String,
+      snapshotFromSpark: Snapshot,
+      strictFileValidation: Boolean = true,
+      ckptVersionExpected: Option[Int] = None,
+      expectV2CheckpointFormat: Boolean = true): Unit = {
+    // Create a snapshot from Spark connector and from kernel.
+    val snapshot = latestSnapshot(path)
+    val snapshotImpl = snapshot.asInstanceOf[SnapshotImpl]
+
+    // Validate metadata/protocol loaded correctly from top-level v2 checkpoint file.
+    val expectedMetadataId =
+      DeltaTable.forPath(path).detail().select("id").collect().head.getString(0)
+    assert(snapshotImpl.getMetadata.getId == expectedMetadataId)
+    assert(snapshotImpl.getProtocol.getMinReaderVersion ==
+      snapshotFromSpark.protocol.minReaderVersion)
+    assert(snapshotImpl.getProtocol.getMinWriterVersion ==
+      snapshotFromSpark.protocol.minWriterVersion)
+    assert(snapshotImpl.getProtocol.getReaderFeatures.asScala.toSet ==
+      snapshotFromSpark.protocol.readerFeatureNames)
+    assert(snapshotImpl.getProtocol.getWriterFeatures.asScala.toSet ==
+      snapshotFromSpark.protocol.writerFeatureNames)
+    assert(snapshot.getVersion(defaultTableClient) == snapshotFromSpark.version)
+
+    // Validate that snapshot read from most recent checkpoint.
+    val expectedV2CkptToRead =
+      ckptVersionExpected.getOrElse(snapshotFromSpark.version - (snapshotFromSpark.version % 2))
```

Comment:
what is this check? `snapshotFromSpark.version % 2`?

Reply:
Checks that it is the most recent even version.